File: src/bulletml_binary/bulletml-converter.cpp
/// BulletML XML to Binary Converter
/**
 * Converts BulletML XML files to binary format (.bmlb)
 * Usage: bulletml-converter input.xml output.bmlb
 */

#include "../bulletml/bulletmlparser-tinyxml.h"
#include "../bulletml/bulletmltree.h"
#include <iostream>
#include <fstream>
#include <map>
#include <vector>
#include <cstring>
#include <cstdint>

class BulletMLBinaryWriter {
public:
    BulletMLBinaryWriter() : current_node_id_(0) {}
    
    bool convert(const std::string& input_xml, const std::string& output_bin) {
        try {
            // Parse XML file
            std::cout << "Parsing XML file: " << input_xml << std::endl;
            BulletMLParserTinyXML parser(input_xml);
            parser.build();
            
            // Prepare for conversion
            buffer_.clear();
            string_table_.clear();
            string_to_id_.clear();
            bullet_refs_.clear();
            action_refs_.clear();
            fire_refs_.clear();
            current_node_id_ = 0;
            node_to_id_.clear();
            
            // Reserve space for header (will write later)
            buffer_.resize(24, 0);
            
            // Write string table placeholder
            uint32_t string_table_offset = buffer_.size();
            
            // Collect all strings and build tree
            std::cout << "Building string table..." << std::endl;
            collectStrings(parser.getBulletRef(0) ? parser.getBulletRef(0)->getParent() : nullptr);
            if (!parser.getBulletRef(0)) {
                // Try to get root from action or fire
                BulletMLNode* root = nullptr;
                if (parser.getTopActions().size() > 0) {
                    root = parser.getTopActions()[0];
                    while (root && root->getParent()) {
                        root = root->getParent();
                    }
                }
                collectStrings(root);
            }
            
            // Collect strings from entire tree starting from root
            BulletMLNode* bulletml = nullptr;
            if (parser.getTopActions().size() > 0) {
                bulletml = parser.getTopActions()[0];
                while (bulletml && bulletml->getParent()) {
                    bulletml = bulletml->getParent();
                }
            }
            
            if (!bulletml) {
                std::cerr << "Error: Could not find bulletml root node" << std::endl;
                return false;
            }
            
            collectStrings(bulletml);
            
            // Write string table
            writeStringTable();
            
            // Write reference maps placeholder
            uint32_t refmap_offset = buffer_.size();
            
            // Build reference maps by traversing tree
            std::cout << "Building reference maps..." << std::endl;
            buildReferenceMaps(bulletml, parser);
            
            // Write reference maps
            writeReferenceMaps();
            
            // Write node tree
            std::cout << "Writing node tree..." << std::endl;
            uint32_t tree_offset = buffer_.size();
            current_node_id_ = 0;
            writeNode(bulletml);
            
            // Write header
            std::cout << "Writing header..." << std::endl;
            writeHeader(parser.isHorizontal(), string_table_offset, refmap_offset, tree_offset);
            
            // Write to file
            std::cout << "Writing to file: " << output_bin << std::endl;
            std::ofstream out(output_bin.c_str(), std::ios::binary);
            if (!out) {
                std::cerr << "Error: Could not open output file" << std::endl;
                return false;
            }
            out.write(reinterpret_cast<const char*>(buffer_.data()), buffer_.size());
            out.close();
            
            std::cout << "Conversion successful!" << std::endl;
            std::cout << "  Input size:  " << getFileSize(input_xml) << " bytes" << std::endl;
            std::cout << "  Output size: " << buffer_.size() << " bytes" << std::endl;
            std::cout << "  Compression: " << (100.0 * buffer_.size() / getFileSize(input_xml)) << "%" << std::endl;
            
            return true;
        }
        catch (const std::exception& e) {
            std::cerr << "Exception during conversion: " << e.what() << std::endl;
            return false;
        }
    }

private:
    void writeUInt8(uint8_t value) {
        buffer_.push_back(value);
    }
    
    void writeUInt16(uint16_t value) {
        buffer_.push_back(value & 0xFF);
        buffer_.push_back((value >> 8) & 0xFF);
    }
    
    void writeUInt32(uint32_t value) {
        buffer_.push_back(value & 0xFF);
        buffer_.push_back((value >> 8) & 0xFF);
        buffer_.push_back((value >> 16) & 0xFF);
        buffer_.push_back((value >> 24) & 0xFF);
    }
    
    void writeUInt32At(size_t offset, uint32_t value) {
        buffer_[offset] = value & 0xFF;
        buffer_[offset + 1] = (value >> 8) & 0xFF;
        buffer_[offset + 2] = (value >> 16) & 0xFF;
        buffer_[offset + 3] = (value >> 24) & 0xFF;
    }
    
    void writeString(const std::string& str) {
        writeUInt16(static_cast<uint16_t>(str.length()));
        buffer_.insert(buffer_.end(), str.begin(), str.end());
    }
    
    uint32_t addString(const std::string& str) {
        if (str.empty()) return 0xFFFFFFFF;
        
        std::map<std::string, uint32_t>::iterator it = string_to_id_.find(str);
        if (it != string_to_id_.end()) {
            return it->second;
        }
        
        uint32_t id = static_cast<uint32_t>(string_table_.size());
        string_table_.push_back(str);
        string_to_id_[str] = id;
        return id;
    }
    
    void collectStrings(BulletMLNode* node) {
        if (!node) return;
        
        // Collect value string
        try {
            // getValue() might throw if no value is set, so we need to check
            // We can't directly check if there's a value, so we use a workaround
            // by checking if it's a node type that typically has values
            BulletMLNode::Name name = node->getName();
            if (name == BulletMLNode::direction || name == BulletMLNode::speed ||
                name == BulletMLNode::wait || name == BulletMLNode::times ||
                name == BulletMLNode::term || name == BulletMLNode::param) {
                // These nodes typically have values - but we need the string form
                // which we don't have direct access to. Skip for now.
            }
        } catch (...) {
            // No value, that's fine
        }
        
        // Recursively collect from children
        BulletMLNode::ChildIterator it;
        for (it = node->childBegin(); it != node->childEnd(); ++it) {
            collectStrings(*it);
        }
    }
    
    void writeStringTable() {
        writeUInt32(static_cast<uint32_t>(string_table_.size()));
        for (size_t i = 0; i < string_table_.size(); ++i) {
            writeString(string_table_[i]);
        }
    }
    
    void buildReferenceMaps(BulletMLNode* node, BulletMLParserTinyXML& parser) {
        // For this implementation, we'll use the parser's internal reference maps
        // This is a simplified version - in a full implementation, we would
        // traverse the tree and build label->id mappings
    }
    
    void writeReferenceMaps() {
        // Write bullet references
        writeUInt32(static_cast<uint32_t>(bullet_refs_.size()));
        for (size_t i = 0; i < bullet_refs_.size(); ++i) {
            writeUInt32(bullet_refs_[i].first);
            writeUInt32(bullet_refs_[i].second);
        }
        
        // Write action references
        writeUInt32(static_cast<uint32_t>(action_refs_.size()));
        for (size_t i = 0; i < action_refs_.size(); ++i) {
            writeUInt32(action_refs_[i].first);
            writeUInt32(action_refs_[i].second);
        }
        
        // Write fire references
        writeUInt32(static_cast<uint32_t>(fire_refs_.size()));
        for (size_t i = 0; i < fire_refs_.size(); ++i) {
            writeUInt32(fire_refs_[i].first);
            writeUInt32(fire_refs_[i].second);
        }
    }
    
    uint8_t nodeTypeToUInt8(BulletMLNode::Name name) {
        return static_cast<uint8_t>(name);
    }
    
    uint8_t valueTypeToUInt8(BulletMLNode::Type type) {
        return static_cast<uint8_t>(type);
    }
    
    void writeNode(BulletMLNode* node) {
        if (!node) return;
        
        uint32_t my_node_id = current_node_id_++;
        node_to_id_[node] = my_node_id;
        
        // Write node header
        writeUInt8(nodeTypeToUInt8(node->getName()));
        writeUInt8(valueTypeToUInt8(node->getType()));
        
        // Count children
        uint16_t child_count = 0;
        BulletMLNode::ChildIterator it;
        for (it = node->childBegin(); it != node->childEnd(); ++it) {
            child_count++;
        }
        writeUInt16(child_count);
        
        // Write ref ID
        BulletMLNode::Name name = node->getName();
        if (name == BulletMLNode::bulletRef || 
            name == BulletMLNode::actionRef || 
            name == BulletMLNode::fireRef) {
            writeUInt32(node->getRefID());
        } else {
            writeUInt32(0xFFFFFFFF);
        }
        
        // Write value string ID (we don't have direct access to the string form)
        writeUInt32(0xFFFFFFFF);
        
        // Write label string ID
        writeUInt32(0xFFFFFFFF);
        
        // Write children
        for (it = node->childBegin(); it != node->childEnd(); ++it) {
            writeNode(*it);
        }
    }
    
    void writeHeader(bool horizontal, uint32_t string_table_offset,
                    uint32_t refmap_offset, uint32_t tree_offset) {
        // Write at beginning
        buffer_[0] = 'B';
        buffer_[1] = 'M';
        buffer_[2] = 'L';
        buffer_[3] = 'B';
        
        writeUInt32At(4, 1); // version
        buffer_[6] = horizontal ? 1 : 0; // orientation
        buffer_[7] = 0; // flags
        
        writeUInt32At(8, string_table_offset);
        writeUInt32At(12, refmap_offset);
        writeUInt32At(16, tree_offset);
        writeUInt32At(20, static_cast<uint32_t>(buffer_.size()));
    }
    
    size_t getFileSize(const std::string& filename) {
        std::ifstream file(filename.c_str(), std::ios::binary | std::ios::ate);
        if (!file) return 0;
        return file.tellg();
    }

private:
    std::vector<uint8_t> buffer_;
    std::vector<std::string> string_table_;
    std::map<std::string, uint32_t> string_to_id_;
    std::vector<std::pair<uint32_t, uint32_t>> bullet_refs_;
    std::vector<std::pair<uint32_t, uint32_t>> action_refs_;
    std::vector<std::pair<uint32_t, uint32_t>> fire_refs_;
    uint32_t current_node_id_;
    std::map<BulletMLNode*, uint32_t> node_to_id_;
};

int main(int argc, char* argv[]) {
    std::cout << "BulletML XML to Binary Converter v1.0" << std::endl;
    std::cout << "=====================================" << std::endl << std::endl;
    
    if (argc < 3) {
        std::cout << "Usage: " << argv[0] << " <input.xml> <output.bmlb>" << std::endl;
        std::cout << std::endl;
        std::cout << "Converts BulletML XML files to binary format." << std::endl;
        std::cout << "Binary files are faster to load and smaller in size." << std::endl;
        return 1;
    }
    
    std::string input = argv[1];
    std::string output = argv[2];
    
    BulletMLBinaryWriter writer;
    if (writer.convert(input, output)) {
        return 0;
    } else {
        std::cerr << "Conversion failed!" << std::endl;
        return 1;
    }
}

File: src/bulletml_binary/bulletml_converter.py
#!/usr/bin/env python3
"""BulletML XML -> BMLB binary converter.

This script converts BulletML XML files into the binary format documented in
src/bulletml_binary/BINARY_FORMAT.md.

Usage:
  python3 bulletml_converter.py input.xml output.bmlb
"""

from __future__ import annotations

import argparse
import io
import struct
import sys
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Optional, Tuple


MAGIC = b"BMLB"
VERSION = 1
NONE_U32 = 0xFFFFFFFF


NODE_TYPE: Dict[str, int] = {
    "bulletml": 0x00,
    "bullet": 0x01,
    "action": 0x02,
    "fire": 0x03,
    "changeDirection": 0x04,
    "changeSpeed": 0x05,
    "accel": 0x06,
    "wait": 0x07,
    "repeat": 0x08,
    "bulletRef": 0x09,
    "actionRef": 0x0A,
    "fireRef": 0x0B,
    "vanish": 0x0C,
    "horizontal": 0x0D,
    "vertical": 0x0E,
    "term": 0x0F,
    "times": 0x10,
    "direction": 0x11,
    "speed": 0x12,
    "param": 0x13,
}

VALUE_TYPE: Dict[str, int] = {
    "none": 0x00,
    "aim": 0x01,
    "absolute": 0x02,
    "relative": 0x03,
    "sequence": 0x04,
}


def _local_name(tag: str) -> str:
    if "}" in tag:
        return tag.split("}", 1)[1]
    return tag


@dataclass
class BinNode:
    name: str
    value_type: int = VALUE_TYPE["none"]
    ref_id: int = NONE_U32
    value_text: Optional[str] = None
    label_text: Optional[str] = None
    children: List["BinNode"] = field(default_factory=list)
    node_id: int = -1


class Converter:
    def __init__(self) -> None:
        self._strings: List[str] = []
        self._string_to_id: Dict[str, int] = {}

        # Domain IDs follow BulletMLParser behavior.
        self._domain_label_to_id: Dict[str, Dict[str, int]] = {
            "bullet": {},
            "action": {},
            "fire": {},
        }

        # Definition node for each domain ref id.
        self._domain_def_nodes: Dict[str, Dict[int, BinNode]] = {
            "bullet": {},
            "action": {},
            "fire": {},
        }

    def convert_file(self, in_path: Path, out_path: Path) -> None:
        root = ET.parse(in_path).getroot()
        root_node, is_horizontal = self._build_tree(root)

        self._assign_node_ids(root_node)
        self._collect_strings(root_node)

        out = io.BytesIO()
        out.write(b"\x00" * 24)  # Header placeholder.

        string_table_offset = out.tell()
        self._write_string_table(out)

        refmap_offset = out.tell()
        self._write_refmaps(out)

        tree_offset = out.tell()
        self._write_node_recursive(out, root_node)

        file_size = out.tell()
        self._write_header(out, is_horizontal, string_table_offset, refmap_offset, tree_offset, file_size)

        out_path.parent.mkdir(parents=True, exist_ok=True)
        out_path.write_bytes(out.getvalue())

    def _build_tree(self, xml_root: ET.Element) -> Tuple[BinNode, bool]:
        root_tag = _local_name(xml_root.tag)
        if root_tag != "bulletml":
            raise ValueError(f"Root element must be <bulletml>, got <{root_tag}>")

        orientation = xml_root.attrib.get("type", "vertical").strip().lower()
        is_horizontal = orientation == "horizontal"

        root_node = self._build_node(xml_root)
        return root_node, is_horizontal

    def _build_node(self, elem: ET.Element) -> BinNode:
        name = _local_name(elem.tag)
        if name not in NODE_TYPE:
            raise ValueError(f"Unsupported BulletML node <{name}>")

        node = BinNode(name=name)

        # Parse attributes matching BulletML parser behavior.
        type_attr = elem.attrib.get("type")
        if type_attr is not None and name != "bulletml":
            t = type_attr.strip()
            if t not in VALUE_TYPE:
                raise ValueError(f"Unsupported type attribute '{t}' on <{name}>")
            node.value_type = VALUE_TYPE[t]

        label_attr = elem.attrib.get("label")
        if label_attr is not None:
            label = label_attr.strip()
            node.label_text = label
            domain = self._domain_for_label(name)
            node.ref_id = self._get_or_create_domain_id(domain, label)
            if name in ("bullet", "action", "fire"):
                self._domain_def_nodes[name][node.ref_id] = node

        # TinyXML parser stores first text node as value for this element.
        if elem.text is not None:
            txt = elem.text.strip()
            if txt:
                node.value_text = txt

        for child in list(elem):
            if not isinstance(child.tag, str):
                continue
            node.children.append(self._build_node(child))

        return node

    @staticmethod
    def _domain_for_label(node_name: str) -> str:
        if node_name == "bulletRef":
            return "bullet"
        if node_name == "actionRef":
            return "action"
        if node_name == "fireRef":
            return "fire"
        return node_name

    def _get_or_create_domain_id(self, domain: str, label: str) -> int:
        if domain not in self._domain_label_to_id:
            # Only bullet/action/fire are valid domains in BulletML refs.
            self._domain_label_to_id[domain] = {}

        domain_map = self._domain_label_to_id[domain]
        if label in domain_map:
            return domain_map[label]

        new_id = len(domain_map)
        domain_map[label] = new_id
        return new_id

    def _assign_node_ids(self, root: BinNode) -> None:
        next_id = 0

        def walk(n: BinNode) -> None:
            nonlocal next_id
            n.node_id = next_id
            next_id += 1
            for c in n.children:
                walk(c)

        walk(root)

    def _add_string(self, s: str) -> int:
        if s in self._string_to_id:
            return self._string_to_id[s]
        idx = len(self._strings)
        self._strings.append(s)
        self._string_to_id[s] = idx
        return idx

    def _collect_strings(self, root: BinNode) -> None:
        def walk(n: BinNode) -> None:
            if n.label_text:
                self._add_string(n.label_text)
            if n.value_text:
                self._add_string(n.value_text)
            for c in n.children:
                walk(c)

        walk(root)

    def _write_header(
        self,
        out: io.BytesIO,
        is_horizontal: bool,
        string_table_offset: int,
        refmap_offset: int,
        tree_offset: int,
        file_size: int,
    ) -> None:
        header = struct.pack(
            "<4sHBBIIII",
            MAGIC,
            VERSION,
            1 if is_horizontal else 0,
            0,
            string_table_offset,
            refmap_offset,
            tree_offset,
            file_size,
        )
        out.seek(0)
        out.write(header)
        out.seek(file_size)

    def _write_string_table(self, out: io.BytesIO) -> None:
        out.write(struct.pack("<I", len(self._strings)))
        for s in self._strings:
            raw = s.encode("utf-8")
            if len(raw) > 0xFFFF:
                raise ValueError(f"String too long for format (max 65535 bytes): {s[:80]}...")
            out.write(struct.pack("<H", len(raw)))
            out.write(raw)

    def _make_ref_entries(self, domain: str) -> List[Tuple[int, int]]:
        label_map = self._domain_label_to_id.get(domain, {})
        def_nodes = self._domain_def_nodes.get(domain, {})

        entries: List[Tuple[int, int]] = []
        for label, ref_id in sorted(label_map.items(), key=lambda kv: kv[1]):
            label_id = self._string_to_id.get(label, NONE_U32)
            node = def_nodes.get(ref_id)
            node_id = node.node_id if node is not None else NONE_U32
            entries.append((label_id, node_id))
        return entries

    def _write_refmaps(self, out: io.BytesIO) -> None:
        for domain in ("bullet", "action", "fire"):
            entries = self._make_ref_entries(domain)
            out.write(struct.pack("<I", len(entries)))
            for label_id, node_id in entries:
                out.write(struct.pack("<II", label_id, node_id))

    def _write_node_recursive(self, out: io.BytesIO, node: BinNode) -> None:
        node_type = NODE_TYPE[node.name]
        value_string_id = NONE_U32
        label_string_id = NONE_U32

        if node.value_text:
            value_string_id = self._string_to_id[node.value_text]
        if node.label_text:
            label_string_id = self._string_to_id[node.label_text]

        out.write(
            struct.pack(
                "<BBHIII",
                node_type,
                node.value_type,
                len(node.children),
                node.ref_id,
                value_string_id,
                label_string_id,
            )
        )

        for child in node.children:
            self._write_node_recursive(out, child)


def _parse_args(argv: List[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Convert BulletML XML files to .bmlb binary format"
    )
    parser.add_argument(
        "input",
        type=Path,
        help="Input XML file or directory containing XML files",
    )
    parser.add_argument(
        "output",
        type=Path,
        help="Output .bmlb file (file mode) or output directory (directory mode)",
    )
    return parser.parse_args(argv)


def main(argv: List[str]) -> int:
    args = _parse_args(argv)

    if not args.input.exists():
        print(f"error: input file not found: {args.input}", file=sys.stderr)
        return 1

    # Directory mode: convert all *.xml files in input directory.
    if args.input.is_dir():
        if args.output.exists() and not args.output.is_dir():
            print(
                f"error: output must be a directory when input is a directory: {args.output}",
                file=sys.stderr,
            )
            return 1

        args.output.mkdir(parents=True, exist_ok=True)
        xml_files = sorted(args.input.glob("*.xml"))
        if not xml_files:
            print(f"error: no XML files found in directory: {args.input}", file=sys.stderr)
            return 1

        success = 0
        failed = 0
        for xml_path in xml_files:
            out_path = args.output / f"{xml_path.stem}.bmlb"
            try:
                conv = Converter()
                conv.convert_file(xml_path, out_path)
                success += 1
                print(f"Converted: {xml_path} -> {out_path}")
            except Exception as exc:  # noqa: BLE001
                failed += 1
                print(f"error: conversion failed for {xml_path}: {exc}", file=sys.stderr)

        print(f"Summary: {success} succeeded, {failed} failed")
        return 0 if failed == 0 else 1

    # Single file mode.
    try:
        conv = Converter()
        conv.convert_file(args.input, args.output)
    except Exception as exc:  # noqa: BLE001
        print(f"error: conversion failed: {exc}", file=sys.stderr)
        return 1

    print(f"Converted: {args.input} -> {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
